cylc/flow/scripts/lint.py

```diff
@@ -0,0 +1,590 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Cylc 728 looks through one or more folders for ``suite*.rc`` files and
+search for Cylc 7 syntax which may be problematic at Cylc 8.
+
+Can be run either as a linter or "in place" (``-i``), leaving comments
+in files. If used in the "in place" mode it is recommended that you ensure
+that you have recorded the state of your workflow in a version control
+system before starting.
+
+.. warning::
+
+   When run with ``-i`` (``--inplace``) mode ``Cylc 728`` changes your files.
+   We strongly recommend committing your workflow to version control
+   before using ``Cylc 728 -i``.
+
+Usage
+^^^^^
+
+.. code-block:: bash
+
+   # run as a linter
+   cylc 728 <paths to workflow directories to check>
+
+   # run inplace
+   cylc 728 --inplace <paths to workflow directories to check>
+   cylc 728 -i <paths to workflow directories to check>
+
+   # Get information about errors:
+   cylc 728 --reference
+   cylc 728 -r
+
+Change Codes
+^^^^^^^^^^^^
+
+"""
+from colorama import Fore
+from optparse import Values
+from pathlib import Path
+import re
+from typing import Generator
+
+from cylc.flow import LOG
+from cylc.flow.option_parsers import (
+    CylcOptionParser as COP,
+)
+from cylc.flow.terminal import cli_function
+
+STYLE_GUIDE = (
+    'https://cylc.github.io/cylc-doc/latest/html/workflow-design-guide/'
+    'style-guide.html#'
+)
+URL_STUB = "https://cylc.github.io/cylc-doc/latest/html/7-to-8/"
+SECTION1 = r'\[\s*{}\s*\]'
+SECTION2 = r'\[\[\s*{}\s*\]\]'
+SECTION3 = r'\[\[\[\s*{}\s*\]\]\]'
+FILEGLOBS = ['*.rc', '*.cylc']
+JINJA2_SHEBANG = '#!jinja2'
+JINJA2_FOUND_WITHOUT_SHEBANG = 'jinja2 found: no shebang (#!jinja2)'
+CHECKS_DESC = {'U': '7 to 8 upgrades', 'S': 'Style'}
+CHECKS = {
+    'U': {
+        re.compile(SECTION1.format('visualization')): {
+            'short': 'section ``[visualization]`` has been removed.',
+            'url': 'summary.html#new-web-and-terminal-uis'
+        },
+        re.compile(SECTION1.format('cylc')): {
+            'short': 'section ``[cylc]`` is now called ``[scheduler]``.',
+            'url': 'summary.html#terminology'
+        },
+        re.compile(SECTION2.format('authentication')): {
+            'short': '``[cylc][authentication]`` is now obsolete.',
+            'url': ''
+        },
+        re.compile(r'^\s*include at start-up\s*='): {
```

Comment:
Can we auto-generate these to avoid writing them out manually?

Reply:
Possibly, but this was a simple attempt. IMO out of scope.